google/cloud/storage/benchmarks/async_throughput_benchmark.cc

```diff
@@ -0,0 +1,298 @@
+  auto handler = [&mu](std::chrono::system_clock::time_point start,
+                       std::string_view name, Result const& r) {
+    std::lock_guard lk(mu);
+    std::cout << FormatTimestamp(start)                                       //
+              << ',' << name                                                  //
+              << ',' << r.object_size                                         //
+              << ',' << r.elapsed.count()                                     //
+              << ','                                                          //
+              << ',' << r.object_name                                         //
+              << ',' << r.generation                                          //
+              << ',' << FormatBandwidthGbPerSecond(r.object_size, r.elapsed)  //
+              << ',' << FormatBandwidthMiBs(r.object_size, r.elapsed)         //
+              << "\n";
+  };
+
+  auto client = gcs_ex::AsyncClient(config.options);
+  auto make_worker = [&] {
+    return std::async(std::launch::async, Worker, client, config, handler);
+  };
+
+  std::cout << "# START\n";
+  std::cout << "Start,Operation,ObjectSize,ElapsedNanos,Unused,ObjectName,"
+               "Generation,Gbps,MiBs\n";
+  std::vector<std::future<void>> workers(config.concurrency);
+  std::generate(workers.begin(), workers.end(), make_worker);
+  for (auto& t : workers) t.get();
+  std::cout << "# DONE\n";
+}
+
+Configuration ParseArgs(std::vector<std::string> argv) {
+  using ::google::cloud::storage_benchmarks::ParseSize;
+  using ::google::cloud::testing_util::OptionDescriptor;
+
+  bool help = false;
+  bool description = false;
+  Configuration config;
+  std::vector<OptionDescriptor> desc{
+      {"--help", "print usage information",
+       [&help](std::string const&) { help = true; }},
+      {"--description", "print benchmark description",
+       [&description](std::string const&) { description = true; }},
+      {"--bucket", "select the bucket for the benchmark",
+       [&config](std::string const& v) { config.bucket = v; }},
+      {"--iterations", "select the number of iterations in the benchmark",
+       [&config](std::string const& v) { config.iterations = std::stoi(v); }},
+      {"--object-size", "select the object size",
+       [&config](std::string const& v) { config.object_size = ParseSize(v); }},
+      {"--grpc-endpoint", "select the endpoint for gRPC",
+       [&config](std::string const& v) {
+         config.options.set<g::EndpointOption>(v);
+       }},
+  };
+  auto usage = google::cloud::testing_util::BuildUsage(desc, argv[0]);
+  google::cloud::testing_util::OptionsParse(desc, std::move(argv));
+  if (help) {
+    std::cerr << usage << "\n";
+    std::exit(0);
+  }
+  if (description) {
+    std::cerr << kDescription << "\n";
+    std::exit(0);
+  }
+  if (config.bucket.empty()) {
+    throw std::invalid_argument("empty value for --bucket option");
+  }
+  return config;
+}
+
+}  // namespace
+
```

Comment:
Nit:  Might it be possible to move the other `main()` here, so they're next to each other?  It would also avoid needing to forward declare `ParseArgs()` and `RunBenchmark()`.

Reply:
Done